src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java

```diff
@@ -10,174 +9,413 @@
+     *     Good alignments offer both high reference uniqueness and read uniqueness, as judged by the requested
+     *     {@code mapQThresholdInclusive} and {@code uniqReadLenInclusive}
+     *     (yes we are doing a hard-filtering but more advanced model is not our priority right now 2017-11-20).
+     * </p>
+     *
+     * Note that "original" is meant to be possibly different from the returned configuration,
+     * but DOES NOT mean the alignments of the contig as given by the aligner,
+     * i.e. the configuration should be one of the best given by
+     * {@link FilterLongReadAlignmentsSAMSpark#pickBestConfigurations(AlignedContig, Set, Double)}.
+     */
+    static Tuple2<List<AlignmentInterval>, List<AlignmentInterval>> removeNonUniqueMappings(final List<AlignmentInterval> originalConfiguration,
+                                                                                            final int mapQThresholdInclusive,
+                                                                                            final int uniqReadLenInclusive) {
+        Utils.validateArg(originalConfiguration.size() > 2,
+                "assumption that input configuration to be fine tuned has more than 2 alignments is violated.\n" +
+                        originalConfiguration.stream().map(AlignmentInterval::toPackedString).collect(Collectors.toList()));
+
+        // two pass, each focusing on removing the alignments of a contig that offers low uniqueness in one sense:
+
+        // first pass is for removing alignments with low REFERENCE UNIQUENESS, using low mapping quality as the criterion
+        final List<AlignmentInterval> selectedAlignments = new ArrayList<>(originalConfiguration.size());
+        final List<AlignmentInterval> lowUniquenessMappings = new ArrayList<>(originalConfiguration.size());
+
+        for (final AlignmentInterval alignment : originalConfiguration) {
+            if (alignment.mapQual >= mapQThresholdInclusive)
+                selectedAlignments.add(alignment);
+            else
+                lowUniquenessMappings.add(alignment);
+        }
+
+        // second pass, the slower one, is to remove alignments offering low READ UNIQUENESS,
+        // i.e. with only a very short part of the read being uniquely explained by this particular alignment;
+        // the steps are:
+        //      search bi-directionally until cannot find overlap any more,
+        //      subtract the overlap from the distance covered on the contig by the alignment.
+        //      This gives unique read region it explains.
+        //      If this unique read region is "short": shorter than {@code uniqReadLenInclusive}), drop it.
+
+        // each alignment has an entry of a tuple2, one for max overlap maxFront, one for max overlap maxRear,
+        // max overlap maxFront is a tuple2 registering the index and overlap bases count
+        final List<MaxOverlapInfo> maxOverlapMap = getMaxOverlapPairs(selectedAlignments);
```

Comment:
I still think this would be nicer as a `Map<AlignmentInterval, MaxOverlapInfo>` -- then you wouldn't need your `k` index variable -- but I won't insist if you like it this way better.

Reply:
done. I wish Java have something similar to the `zip` function in Python....